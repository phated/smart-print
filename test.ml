(** A small program to test the library on different samples.
    Run with [make test]. Compare its output to the reference in [test.out]. *)
open SmartPrint

(** The syntax tree: variables, applications, functions, let-definitions, tuples. *)
type t =
  | Var of string
  | App of t * t
  | Fun of string * t
  | Let of string * t * t
  | Tuple of t list

(** Pretty-print an expression, enclosing it in parenthesis if [paren] flag is set. *)
let rec pp (paren : bool) (e : t) : SmartPrint.t =
  let if_parens d =
    if paren then parens d else d in
  match e with
  | Var x -> !^ x
  | App (e1, e2) -> nest @@ if_parens (pp true e1 ^^ pp true e2)
  | Fun (x, e) -> nest @@ parens (!^ "fun" ^^ !^ x ^^ !^ "->" ^^ pp false e)
  | Let (x, e1, e2) ->
    nest (!^ "let" ^^ !^ x ^^ !^ "=" ^^ pp false e1 ^^ !^ "in" ^^ hardline ^^ pp false e2)
  | Tuple es -> OCaml.list (pp false) es

(** A sample of expressions. *)
let es = [
  App (Var "f", Var "x");
  App (Var "fdsgoklkmeee", Var "xdsgsdg");
  Fun ("x", App (Var "fdsgo", App (Var "x", Var "xdsgsdg")));
  Fun ("x", App (Var "fdsgvbcvvfo", App (Var "xffd", Var "xdsgsdg")));
  Tuple [];
  Tuple [Var "x"; Var "y"];
  Tuple (List.map (fun x -> Var x) ["kjh"; "lj"; "iop"; "rt"; "vbn"; "hjk"; "gkgytuuhi"]);
  Let ("x", Var "x", Var "y");
  Let ("x", Fun ("x", App (Var "fdsgo", App (Var "x", Var "xdsgsdg"))), Var "y")]

(** Display the contents of a document and its AST. *)
let print_document (d : SmartPrint.t) : unit =
  let width = 25 in
  let tab = 2 in
  to_stdout 160 tab @@ Debug.pp_document_after_rendering width tab d;
  print_newline ();
  print_endline @@ String.make width '*';
  print_endline @@ to_string width tab d;
  print_endline @@ String.make width '*'

(** The main function. *)
let main () =
  print_document (!^ "hello" ^^ !^ "world" ^^ hardline ^^ nest (!^ "gre" ^^ nest_all (!^ "arg" ^^ !^ "arg")));
  es |> List.iter (fun e ->
    print_document @@ pp false e);
  print_document @@ sub_string "hfgs_kjl_df" 5 3;
  print_document @@ words "Lorem     ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
  print_document @@ lines "adipisicing elit,\nsed do eiusmod tempor\nincididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud ";
  print_document @@ nest @@ lines "adipisicing elit,\n\nsed do eiusmod tempo incididunt ut labore et dolore";
  print_document @@ indent @@ nest @@ lines "adipisicing elit,\n\nsed do eiusmod tempo incididunt ut labore et dolore";
  print_document @@ indent @@ pp false @@ Let ("x", Var "x", Let ("x", Var "x", Var "y"));
  print_document @@ pp false @@ App (Var "f", Let ("x", Var "x", Var "y"));
  print_document @@ OCaml.unit () ^^ OCaml.tuple [!^ "x"; OCaml.int 0]

;;main ()
