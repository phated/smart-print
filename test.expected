GroupOne (false, "hello", Space, "world", Space, Hardline, GroupOne (true, "gre", Space, GroupAll (true, "arg", Space, "arg")))
*************************
hello world
gre arg arg
*************************
GroupOne (false, GroupOne (true, "f", Space, "x"))
*************************
f x
*************************
GroupOne (false, GroupOne (true, "fdsgoklkmeee", Space, "xdsgsdg"))
*************************
fdsgoklkmeee xdsgsdg
*************************
GroupOne (false,
  GroupOne (true,
    "(",
    "fun",
    Space,
    "x",
    Space,
    "->",
    Hardline,
    Indent (1, GroupOne (false, GroupOne (true, "fdsgo", Space, GroupOne (true, "(", "x", Space, "xdsgsdg", ")")), ")"))))
*************************
(fun x ->
  fdsgo (x xdsgsdg))
*************************
GroupOne (false,
  GroupOne (true,
    "(",
    "fun",
    Space,
    "x",
    Space,
    "->",
    Hardline,
    Indent (1, GroupOne (false, GroupOne (true, "fdsgvbcvvfo", Hardline, Indent (1, GroupOne (false, GroupOne (true, "(", "xffd", Space, "xdsgsdg", ")")))), ")"))))
*************************
(fun x ->
  fdsgvbcvvfo
    (xffd xdsgsdg))
*************************
GroupOne (false, "[", GroupAll (true, Space, Space, Space, Space), "]")
*************************
[ ]
*************************
GroupOne (false, "[", GroupAll (true, Space, Space, "x", ";", Space, Space, "y", Space, Space), "]")
*************************
[ x; y ]
*************************
GroupOne (false,
  "[",
  GroupAll (true,
    Hardline,
    Indent (1,
      GroupAll (false,
        "kjh",
        ";",
        Hardline,
        "lj",
        ";",
        Hardline,
        "iop",
        ";",
        Hardline,
        "rt",
        ";",
        Hardline,
        "vbn",
        ";",
        Hardline,
        "hjk",
        ";",
        Hardline,
        "gkgytuuhi",
        Hardline))),
  "]")
*************************
[
  kjh;
  lj;
  iop;
  rt;
  vbn;
  hjk;
  gkgytuuhi
]
*************************
GroupOne (false, GroupOne (true, "let", Space, "x", Space, "=", Space, "x", Space, "in", Space, Hardline, "y"))
*************************
let x = x in
y
*************************
GroupOne (false,
  GroupOne (true,
    "let",
    Space,
    "x",
    Space,
    "=",
    Hardline,
    Indent (1,
      GroupOne (false,
        GroupOne (true,
          "(",
          "fun",
          Space,
          "x",
          Space,
          "->",
          Hardline,
          Indent (1, GroupOne (false, GroupOne (true, "fdsgo", Space, GroupOne (true, "(", "x", Space, "xdsgsdg", ")")), ")"))),
        Space,
        "in",
        Space,
        Hardline,
        Indent (-1, GroupOne (false, "y"))))))
*************************
let x =
  (fun x ->
    fdsgo (x xdsgsdg)) in
y
*************************
GroupOne (false, "kjl")
*************************
kjl
*************************
GroupOne (false,
  GroupOne (false,
    "Lorem",
    Space,
    "ipsum",
    Space,
    "dolor",
    Space,
    "sit",
    Hardline,
    "amet,",
    Space,
    "consectetur",
    Hardline,
    "adipisicing",
    Space,
    "elit,",
    Space,
    "sed",
    Space,
    "do",
    Hardline,
    "eiusmod",
    Space,
    "tempor",
    Space,
    "incididunt",
    Hardline,
    "ut",
    Space,
    "labore",
    Space,
    "et",
    Space,
    "dolore",
    Space,
    "magna",
    Hardline,
    "aliqua.",
    Space,
    "Ut",
    Space,
    "enim",
    Space,
    "ad",
    Space,
    "minim",
    Hardline,
    "veniam,",
    Space,
    "quis",
    Space,
    "nostrud",
    Hardline,
    "exercitation",
    Space,
    "ullamco",
    Hardline,
    "laboris",
    Space,
    "nisi",
    Space,
    "ut",
    Space,
    "aliquip",
    Hardline,
    "ex",
    Space,
    "ea",
    Space,
    "commodo",
    Space,
    "consequat.",
    Hardline,
    "Duis",
    Space,
    "aute",
    Space,
    "irure",
    Space,
    "dolor",
    Space,
    "in",
    Hardline,
    "reprehenderit",
    Space,
    "in",
    Hardline,
    "voluptate",
    Space,
    "velit",
    Space,
    "esse",
    Hardline,
    "cillum",
    Space,
    "dolore",
    Space,
    "eu",
    Space,
    "fugiat",
    Hardline,
    "nulla",
    Space,
    "pariatur.",
    Space,
    "Excepteur",
    Hardline,
    "sint",
    Space,
    "occaecat",
    Space,
    "cupidatat",
    Hardline,
    "non",
    Space,
    "proident,",
    Space,
    "sunt",
    Space,
    "in",
    Hardline,
    "culpa",
    Space,
    "qui",
    Space,
    "officia",
    Hardline,
    "deserunt",
    Space,
    "mollit",
    Space,
    "anim",
    Space,
    "id",
    Hardline,
    "est",
    Space,
    "laborum."))
*************************
Lorem ipsum dolor sit
amet, consectetur
adipisicing elit, sed do
eiusmod tempor incididunt
ut labore et dolore magna
aliqua. Ut enim ad minim
veniam, quis nostrud
exercitation ullamco
laboris nisi ut aliquip
ex ea commodo consequat.
Duis aute irure dolor in
reprehenderit in
voluptate velit esse
cillum dolore eu fugiat
nulla pariatur. Excepteur
sint occaecat cupidatat
non proident, sunt in
culpa qui officia
deserunt mollit anim id
est laborum.
*************************
GroupOne (false,
  "adipisicing elit,", Hardline, "sed do eiusmod tempor", Hardline, "incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud ")
*************************
adipisicing elit,
sed do eiusmod tempor
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud 
*************************
GroupOne (false, GroupOne (true, "adipisicing elit,", Hardline, Hardline, "sed do eiusmod tempo incididunt ut labore et dolore"))
*************************
adipisicing elit,

sed do eiusmod tempo incididunt ut labore et dolore
*************************
GroupOne (false, Indent (1, GroupOne (true, "adipisicing elit,", Hardline, Hardline, "sed do eiusmod tempo incididunt ut labore et dolore")))
*************************
  adipisicing elit,
  
  sed do eiusmod tempo incididunt ut labore et dolore
*************************
GroupOne (false,
  Indent (1,
    GroupOne (true,
      "let",
      Space,
      "x",
      Space,
      "=",
      Space,
      "x",
      Space,
      "in",
      Space,
      Hardline,
      GroupOne (true, "let", Space, "x", Space, "=", Space, "x", Space, "in", Space, Hardline, "y"))))
*************************
  let x = x in
  let x = x in
  y
*************************
GroupOne (false,
  GroupOne (true, "f", Hardline, Indent (1, GroupOne (false, GroupOne (true, "let", Space, "x", Space, "=", Space, "x", Space, "in", Space, Hardline, "y")))))
*************************
f
  let x = x in
  y
*************************
GroupOne (false, "()", Space, "(", GroupOne (true, "x", ",", Space, Space, "0"), ")")
*************************
() (x, 0)
*************************
